examples/temporal_graph_learning/gnn_encoder.py

```diff
@@ -0,0 +1,218 @@
+    g: DistGraph
+        The distributed graph
+    h_dim : int
+        Hidden dimension
+    out_dim : int
+        Output dimension
+    num_hidden_layers : int
+        Number of hidden layers. Total GNN layers is equal to num_hidden_layers + 1. Default 1
+    num_heads : int
+        Number of attention heads
+    dropout : float
+        Dropout. Default 0.
+    """
+    def __init__(self, g, h_dim, out_dim, num_hidden_layers, num_heads=0, dropout=0):
+        super(TemporalRelationalGraphEncoder, self).__init__(h_dim, out_dim, num_hidden_layers)
+        self.node_fields = get_unique_nfields(g.etypes)
+        for _ in range(num_hidden_layers):
+            self.layers.append(
+                ResidualTemporalRelationalGraphConv(
+                    g,
+                    in_feat=h_dim,
+                    out_feat=h_dim,
+                    num_heads=num_heads,
+                    dropout=dropout,
+                    activation=nn.LeakyReLU(),
+                )
+            )
+        self.layers.append(
+            ResidualTemporalRelationalGraphConv(
+                g,
+                in_feat=h_dim,
+                out_feat=out_dim,
+                num_heads=num_heads,
+                dropout=dropout,
+                activation=nn.LeakyReLU(),
+            )
+        )
+
+    def forward(self, blocks, inputs):
+        """Forward computation.
+
+        Parameters
+        ----------
+        blocks: DGL MFGs
+            Sampled subgraph in DGL MFG
+        inputs: dict[str, torch.Tensor]
+            Input node feature for each node type.
+        """
+        h = to_per_field_nfeats(inputs, self.node_fields)
+        for layer, block in zip(self.layers, blocks):
+            h = layer(block, h)
+        h = average_over_fields(h)
+        return h
+
+class ResidualTemporalRelationalGraphConv(nn.Module):
+    """ Relational temporal graph conv layer with residual connection.
+
+    Parameters
+    ----------
+    g: DistGraph
+        The distributed graph
+    in_feat : int
+        Input dimension
+    out_dim : int
+        Output dimension
+    num_heads : int
+        Number of attention heads
+    activation: torch.nn.Module
+        Activation function.
+    dropout : float
+        Dropout. Default 0.
+    """
+    def __init__(self, g, in_feat, out_feat, num_heads, activation=None, dropout=0.0):
+        super(ResidualTemporalRelationalGraphConv, self).__init__()
+        self.in_feat = in_feat
+        self.out_feat = out_feat
+        self.activation = activation
+        self.dropout = nn.Dropout(dropout)
+
+        self.conv = TemporalRelationalGraphConv(g, in_feat, out_feat, num_heads)
+        self.loop_weights = nn.Linear(in_feat, out_feat, bias=True)
+
+    def forward(self, g, inputs):
+        """Forward computation. Residual connection is used.
+
+        Parameters
+        ----------
+        g: DGL MFGs
+            Sampled subgraph in DGL MFG
+        inputs: dict[str, torch.Tensor]
```

Comment:
is this true? i thought `inputs` is a dict of dicts of tensors.

Reply:
Fixed. Change to dict[str, dict[str, torch.Tensor]].